File: DSA/LC/355_design_twitter.cpp
#include <vector>
#include <unordered_map>
#include <queue>

constexpr int FEED_MAX{10};

struct User;
struct Tweet;

struct Tweet
{
  static int shared_time;
  User *owner{};
  int id{};
  int createdAt{shared_time++};

  Tweet() = default;

  Tweet(const int id, User *const owner) : id{id}, owner{owner}
  {
  }

  constexpr bool operator<(const Tweet &other) const
  {
    return createdAt < other.createdAt;
  }
  constexpr bool operator>(const Tweet &other) const
  {
    return createdAt > other.createdAt;
  }
};

int Tweet::shared_time{0};

struct User
{
  std::vector<Tweet> tweets{};
  std::vector<int> feed{};
  std::unordered_map<int, User *> followers{};
  std::unordered_map<int, User *> following{};
  int id{};

  User() = default;

  User(const int id) : id{id}
  {
  }

  void postTweet(int tweetId)
  {
    const Tweet tweet{tweetId, this};
    tweets.push_back(tweet);
  }

  void addFollower(User *const newFollower)
  {
    followers[newFollower->id] = newFollower;
  }

  void follow(User *const userToFollow)
  {
    if (userToFollow == this)
      return;
    following[userToFollow->id] = userToFollow;
    userToFollow->addFollower(this);
  }

  void removeFollower(User *const follower)
  {
    followers.erase(follower->id);
  }

  void unfollow(User *const userToUnfollow)
  {
    if (userToUnfollow == this)
      return;
    following.erase(userToUnfollow->id);
    userToUnfollow->removeFollower(this);
  }

  void updateFeed()
  {
    feed.clear();
    std::priority_queue<Tweet> firstTweetFromEach{};
    std::unordered_map<User *, int> users_tweetIndex{};

    if (!tweets.empty())
    {
      users_tweetIndex[this] = static_cast<int>(tweets.size()) - 1;
      firstTweetFromEach.push(tweets[tweets.size() - 1]);
    }

    for (const auto &[followeeId, followee] : following)
    {
      if (!followee->tweets.empty())
      {
        users_tweetIndex[followee] = static_cast<int>(followee->tweets.size()) - 1;
        firstTweetFromEach.push(followee->tweets[users_tweetIndex[followee]]);
      }
    }

    while (!firstTweetFromEach.empty() && feed.size() < FEED_MAX)
    {
      Tweet newestTweet = firstTweetFromEach.top();
      feed.push_back(newestTweet.id);
      firstTweetFromEach.pop();
      if (--users_tweetIndex[newestTweet.owner] >= 0)
      {
        firstTweetFromEach.push(newestTweet.owner->tweets[users_tweetIndex[newestTweet.owner]]);
      }
    }
  }
};

class Twitter
{
  std::unordered_map<int, User> users{};

public:
  Twitter()
  {
  }

  void postTweet(int userId, int tweetId)
  {
    if (users.find(userId) == users.end())
    {
      users.emplace(userId, User(userId));
    }
    users[userId].postTweet(tweetId);
  }

  std::vector<int> getNewsFeed(int userId)
  {
    users[userId].updateFeed();
    return users[userId].feed;
  }

  void follow(int followerId, int followeeId)
  {
    if (users.find(followerId) == users.end())
    {
      users.emplace(followerId, User(followerId));
    }
    if (users.find(followeeId) == users.end())
    {
      users.emplace(followeeId, User(followeeId));
    }

    users[followerId].follow(&users[followeeId]);
  }

  void unfollow(int followerId, int followeeId)
  {
    if (users.find(followerId) == users.end())
    {
      users.emplace(followerId, User(followerId));
    }
    if (users.find(followeeId) == users.end())
    {
      users.emplace(followeeId, User(followeeId));
    }

    users[followerId].unfollow(&users[followeeId]);
  }
};

File: DSA/LC/355_design_twitter.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

class Twitter {

  private int count;
  private Map<Integer, List<int[]>> tweetMap;
  private Map<Integer, Set<Integer>> followMap;

  public Twitter() {
    count = 0;
    tweetMap = new HashMap<>();
    followMap = new HashMap<>();
  }

  public void postTweet(int userId, int tweetId) {
    tweetMap.computeIfAbsent(userId, k -> new ArrayList<>()).add(new int[] { count--, tweetId });
  }

  public List<Integer> getNewsFeed(int userId) {
    List<Integer> res = new ArrayList<>();
    PriorityQueue<int[]> minHeap = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));

    followMap.computeIfAbsent(userId, k -> new HashSet<>()).add(userId);
    for (int followeeId : followMap.get(userId)) {
      if (tweetMap.containsKey(followeeId)) {
        List<int[]> tweets = tweetMap.get(followeeId);
        int index = tweets.size() - 1;
        int[] tweet = tweets.get(index);
        minHeap.offer(new int[] { tweet[0], tweet[1], followeeId, index });
      }
    }

    while (!minHeap.isEmpty() && res.size() < 10) {
      int[] curr = minHeap.poll();
      res.add(curr[1]);
      int index = curr[3];
      if (index > 0) {
        int[] tweet = tweetMap.get(curr[2]).get(index - 1);
        minHeap.offer(new int[] { tweet[0], tweet[1], curr[2], index - 1 });
      }
    }
    return res;
  }

  public void follow(int followerId, int followeeId) {
    followMap.computeIfAbsent(followerId, k -> new HashSet<>()).add(followeeId);
  }

  public void unfollow(int followerId, int followeeId) {
    followMap.computeIfPresent(followerId, (k, v) -> {
      v.remove(followeeId);
      return v;
    });
  }
}
